### PR TITLE
[FLINK-17445][State Processor] Add Scala support for OperatorTransformation

### PR DESCRIPTION
**Pull request by brandonbevans**
## What is the purpose of the change
Currently there's no way to use a Scala DataSet with the OperatorTransform object. This PR adds support for the Scala DataSet in this part of the state processor API.

## Brief change log

*(for example:)*
  - *Created new scala module, flink-state-processing-api-scala*
  - *Mirrored OperatorTransform's bootstrapWith method to accept a Scala DataSet*

## Verifying this change

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): yes
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-examples/flink-examples-streaming/pom.xml [hunk 1]
```diff
@@ -37,6 +37,17 @@ under the License.
...
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-state-processor-api_${scala.binary.version}</artifactId>
+			<version>${project.version}</version>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-state-processor-api-scala_${scala.binary.version}</artifactId>
+			<version>${project.version}</version>
+		</dependency>
```
Comment (sjwiesman) [inline flink-examples/flink-examples-streaming/pom.xml:50]: Unrelated changes, please remove

File: flink-libraries/flink-state-processing-api-scala/pom.xml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+			<scope>test</scope>
+		</dependency>
+
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-runtime_${scala.binary.version}</artifactId>
+			<version>${project.version}</version>
+			<type>test-jar</type>
+			<scope>test</scope>
+		</dependency>
```
Comment (sjwiesman) [inline flink-libraries/flink-state-processing-api-scala/pom.xml:99]: It doesn't look like you are using any of these test dependencies, please remove.

File: flink-libraries/flink-state-processing-api-scala/pom.xml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+				<executions>
+					<execution>
+						<goals>
+							<goal>test-jar</goal>
+						</goals>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+	</build>
```
Comment (sjwiesman) [inline flink-libraries/flink-state-processing-api-scala/pom.xml:117]: What is this plugin doing?

File: flink-libraries/flink-state-processing-api-scala/pom.xml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+	</parent>
+
+	<artifactId>flink-state-processor-api-scala_${scala.binary.version}</artifactId>
+	<name>flink-state-processor-api-scala</name>
+
+	<packaging>jar</packaging>
+
+	<dependencies>
+
+		<!-- core dependencies -->
```
Comment (sjwiesman) [inline flink-libraries/flink-state-processing-api-scala/pom.xml:40]: There should be an explicit `provided` dependency on `flink-scala`.

File: flink-libraries/flink-state-processing-api-scala/src/test/scala/org/apache/flink/state/api/scala/OperatorTransformationTest.scala [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+import org.apache.flink.api.scala.{ExecutionEnvironment, _}
+import org.apache.flink.util.TestLogger
+import org.junit.Test
+
+class OperatorTransformationTest extends TestLogger {
+
+  @Test
+  @throws[Exception]
+  def testScalaDatasetBootstrapWith() {
```
Comment (sjwiesman) [inline flink-libraries/flink-state-processing-api-scala/src/test/scala/org/apache/flink/state/api/scala/OperatorTransformationTest.scala:11]: Is this test just verifying that it compiles? It doesn't execute any code.

File: flink-libraries/flink-state-processing-api-scala/src/main/scala/org/apache/flink/state/api/scala/OperatorTransformation.scala [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package org.apache.flink.state.api.scala
+
+import org.apache.flink.api.scala.DataSet
+import org.apache.flink.state.api.{OneInputOperatorTransformation, OperatorTransformation => JOperatorTransformation}
+
+object OperatorTransformation {
+  def bootstrapWith[T](dataSet: DataSet[T]): OneInputOperatorTransformation[T] = {
+    JOperatorTransformation.bootstrapWith(dataSet.javaSet)
```
Comment (sjwiesman) [inline flink-libraries/flink-state-processing-api-scala/src/main/scala/org/apache/flink/state/api/scala/OperatorTransformation.scala:8]: Unfortunately this isn't going to work. See my main comment.

**Comment by sjwiesman** [comment]
Also, the build is failing due to missing license headers and checkstyle issues. If you run `mvn verify` from within `flink-state-processing-api-scala` locally you can check these issues.

**Comment by brandonbevans** [comment]
@sjwiesman Thank you for the thorough review I really appreciate it. I'll make the advised changes / clean up and get back.

**Comment by zentol** [comment]
Scala APIs are deprecated and will not be further extended.
